File: conformance/tests/go-dsl-bytestring-literal/go-dsl-bytestring-literal.runar.go
//go:build ignore

package contract

import "runar"

// Exercises two Go-DSL features at once:
//   1. `runar.BigintBig` declared as a property type — the DSL parser maps
//      this to the same `bigint` primitive as `runar.Bigint`, so arithmetic
//      / comparison works against ordinary Bigint values.
//   2. `runar.ByteString("literal")` as an inline byte-string literal — the
//      DSL parser decodes the Go string escape sequences (\x00, \x6a) and
//      emits a ByteString literal whose hex value represents the raw bytes.
type GoDslBytestringLiteral struct {
	runar.SmartContract
	Target   runar.BigintBig  `runar:"readonly"`
	Expected runar.ByteString `runar:"readonly"`
}

func (c *GoDslBytestringLiteral) Check(a runar.Bigint, b runar.BigintBig) {
	runar.Assert(a+b == c.Target)
	runar.Assert(runar.ByteString("\x00\x6a") == c.Expected)
}

File: examples/java/src/main/java/runar/examples/go-dsl-bytestring-literal/GoDslBytestringLiteral.runar.java
package runar.examples.godslbytestringliteral;

import runar.lang.SmartContract;
import runar.lang.annotations.Public;
import runar.lang.annotations.Readonly;
import runar.lang.types.Bigint;
import runar.lang.types.ByteString;

import static runar.lang.Builtins.assertThat;

/**
 * GoDslBytestringLiteral -- mirrors
 * {@code examples/go/go-dsl-bytestring-literal/GoDslBytestringLiteral.runar.go}.
 *
 * <p>The Go-DSL fixture exercises Go-only surface features
 * ({@code runar.BigintBig} property type, {@code runar.ByteString("\\x00\\x6a")}
 * literal). In Java the equivalent surface is plain {@link Bigint} /
 * {@link ByteString} plus {@code ByteString.fromHex("006a")}. Both lower to the
 * same primitives in the AST.
 */
class GoDslBytestringLiteral extends SmartContract {

    @Readonly Bigint target;
    @Readonly ByteString expected;

    GoDslBytestringLiteral(Bigint target, ByteString expected) {
        super(target, expected);
        this.target = target;
        this.expected = expected;
    }

    @Public
    void check(Bigint a, Bigint b) {
        assertThat(a.plus(b).equals(target));
        assertThat(ByteString.fromHex("006a").equals(expected));
    }
}
